### ISSUE TITLE
Support DenyAssignments for MG-scoped Deployment Stacks

### ISSUE THREAD

**Issue by alex-frankel** [Azure/deployment-stacks#137]

**Comment by marsontret** [comment]
Covered by to #156?

**Comment by snarkywolverine** [comment]
@marsontret No, #156 only adds Management Groups to the list of managed resources. Creating a deny assignment on those management groups is still not a supported scenario, which is why we are asking customers to upvote this if they have a scenario where this would benefit them.

**Comment by marsontret** [comment]
Got it. Thanks.

**Comment by sbebrys** [comment]
I have one case which needs deny assignments on MG level. 

I wanna provide via stack Azure Polices where **definitions** of policies and initiatives will be located on mg level (for easier sharing that to each sub or rg) but policy **assignments** on sub or rg level.

Goal is also to provide also deny assignments to make this setup readonly and be managed only via code. 

For now I cannot do that even I split definitions (without deny) and assignments (with deny) to different stacks, because assignments reference to policy definition from mg level so I need choose this scope for deployment as least.

**Comment by snarkywolverine** [comment]
Hi @sbebrys -

I believe the following should work for your scenario:
1. Split the policy definitions and policy assignments.
2. Create an MG-scoped stack for the policy assignments, but set the deployment scope to a subscription.  You should be able to specify denySettings in this case.  Note: The deployment stack must be at the management group that is the direct parent of the target subscription.
3. The template should directly contain the policy assignments (referencing the policy definitions at management group scope).  This will be deployed to the subscription and protected by deny assignments, and subscription owners won't have access to the management group to modify the deployment stack.

Does that make sense? Let me know if you have any questions.

**Comment by sbebrys** [comment]
Hi @snarkywolverine,

I checked that in smaller example, I tried create just role assignment on subscription level but stack for it on upper MG level and I got error:

![image](https://github.com/Azure/deployment-stacks/assets/22738365/74b0e068-34d9-46b3-8b93-9d70a4ef3c1f)

So it looks that, for deny settings, api try challenge not resource deployment scope, but targetScope of whole deployment :|

**Comment by snarkywolverine** [comment]
@sbebrys It looks like you aren't deploying to subscription scope.  You should pass --deployment-subscription <subID> as an additional parameter.

**Comment by sbebrys** [comment]
@snarkywolverine great I miss that, thanks. 

Finally It could works for me but only partial. I could push policy assignments to read-only state but not policy definitions itself, so still someone could change policy and its behavior outside assignments. Deny assignments on mg level could be helpful in this case.

**Comment by humanascode** [comment]
Voting for this one, is there a plan to implement this functionality? This would be beneficial for cases where you want to protect your policy assignments and definition at the MG level

**Comment by pbo-top** [comment]
We would like to protect Subscription Budgets to be modded/deleted by its Contributers and Owners. DenyAssignments at MG level could achive just that. Upvoted.

**Comment by paul-mccormack** [comment]
I would like to upvote this.  I want to deploy policy assignments at the MG to require cost tracking and ownership tags on RG's and inherit to resources.  These are currently assigned to subscriptions but as new subscriptions are added the policies aren't being assigned.  Deploying this as a stack to the parent MG would make this automatic.  I want to deploy the assignments via Azure DevOps pipeline and prevent modification and deletion of the assignments from the portal.  Hope that makes sense.  Thanks.

**Comment by snarkywolverine** [comment]
@pbo-top Do the budgets exist at management group scope, or are they saved in the subscription?

If they're saved in the subscription, you should be able to achieve this scenario today with the following Pwsh command:
New-AzManagementGroupDeploymentStack ... -DeploymentSubscriptionId <subId> -DenySettings ...

This creates the deployment stack at MG scope, but deploys directly to the specified subscription ID.  In this case, DenySettings are supported.  Does this address your scenario?

@paul-mccormack - will this solution work for your scenario as well?  Though note that tags are not currently supported as a denyable resource. See #151 for details.

**Comment by paul-mccormack** [comment]
@snarkywolverine - Hi,  Thanks for the reply.  I would need to test but from my understanding I would need to create a stack for every subscription under the management group and for future subscriptions that might be created?

I'm not looking to protect the tags with a deny, it's the policy assignments I want to protect.  There are a bunch of other policy assignments included in this task.  Things like allowed regions, sku's, etc.  I want to run this from DevOps with only the connector service principal and the break glass account being excluded from the deny.  Effectively blocking anyone in the portal from removing the policy assignment and deploying something in a disallowed region, or some other thing we want to block or control.

As an example, and please let me know if I am misunderstanding.  Say I have MG1 which contains sub1, sub2 and sub3.  I would need to deploy three stacks at MG1 with DeploymentSubscriptionId sub1, sub and sub3 respectively.  Later someone creates sub4 and sub5.  Requiring another two stacks to be created to target them.

If that is the case and I want to ensure these policy assignments are automated for every current and future subscription, I could look at putting a workflow in place for creating new subs using DevOps then running this deployment as part of the workflow. But if the deny assignments inherit from the MG it would be a lot less work.

**Comment by snarkywolverine** [comment]
@paul-mccormack Thanks for clarifying the scenario. I'm guessing there are users who have access to the management group who shouldn't be able to modify the policy, which is why the deny assignment is valuable?  If so, what permissions do the users have in the MG?  Also, are you more concerned about malicious policy changes, or accidental ones?

Your answers can help us guide the design and figure out how easy it is to potentially implement something like this.

**Comment by paul-mccormack** [comment]
@snarkywolverine I guess the goal would be to control how the policies are modified, using Inf as Code to maintain a record and gain approval for the modifications.  Not all of our technical teams are at the same level with Bicep deployments.  We regularly find modifications happen in the portal with the good intention of retrospective code or documentation updates, which unfortunately sometimes get forgotten about.

So, potentially people who can elevate up to Owner role would need to be blocked by the deny assignment.  I think the approval mechanism gained by using a CI/CD pipeline would effectively control both malicious and accidental changes at the same time.  Hope that helps.  Thanks.

**Comment by pbo-top** [comment]
@snarkywolverine , sorry for a delayed answer.
> @pbo-top Do the budgets exist at management group scope, or are they saved in the subscription?
> 

Thanks for the suggestion, I'm playing with this using the az-cli and Bicep though, but it is actually what I'm trying to do - perform a Management Group deployment containing module with a Budget on a Subscription. But that generates this err:
```
{"code": "DeploymentStackInvalidDeploymentStackDefinition", "message": "The deployment stack 'test-budget-stack-001' is invalid, because deny assignments are only supported for deployments at subscription scope."}
```
Command:
```
az stack mg create `
--name 'test-budget-stack-001' `
--location 'swedencentral' `
--management-group-id 'deploy-stack-test-001' `
--template-file 'mg-wrapper.bicep' `
--parameters subscriptionID='234234-23234...`
--action-on-unmanage 'deleteResources' `
--deny-settings-mode 'denyWriteAndDelete' `
--deny-settings-excluded-principals '1231-123123-123123...' `
--deny-settings-apply-to-child-scopes
```
mg-wrapper.bicep:
```
targetScope = 'managementGroup'

param subscriptionID string

module subDeployModule 'budget1.bicep' = {
  name: 'deployToSub'
  scope: subscription(subscriptionID)
}
```
The budget1.bicep is just a Budget resource at Subscription targetScope.

If I omit the "deny-settings" from the cli call the stack is rolled nicely on with a budget.

**Comment by MichalZezula** [comment]
@pbo-top have you been able to figure it out ?
We have very siliar use case where I need to make sure the stack is deployed on the MG level where users cannot touch it, but I need to actually deploy resources on the subscription level.
Deployment part is not a problem but no matter what, I cannot make the deny settings to work.

**Comment by snarkywolverine** [comment]
@pbo-top @MichalZezula Deployment Stacks absolutely supports this scenario!

The key is to skip the 'wrapper' bicep file, and instead deploy the subscription template directly, but using the DeploymentSubscription parameter.



PS: https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2372453056
CLI: https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2067033907

So the cmd for @pbo-top would be something like the following (adding a line break and rearranging parameters for clarity)

````
az stack mg create `
--name 'test-budget-stack-001' `
--location 'swedencentral' `
--management-group-id 'deploy-stack-test-001' `
--parameters subscriptionID='234234-23234...`
--action-on-unmanage 'deleteResources' `
--deny-settings-mode 'denyWriteAndDelete' `
--deny-settings-excluded-principals '1231-123123-123123...' `
--deny-settings-apply-to-child-scopes

--template-file 'budget1.bicep' ` <<< ***CHANGED***
--deployment-subscription 248248-....  <<< ***ADDED***
````

**Comment by MichalZezula** [comment]
> [@pbo-top](https://github.com/pbo-top) [@MichalZezula](https://github.com/MichalZezula) Deployment Stacks absolutely supports this scenario!
> 
> The key is to skip the 'wrapper' bicep file, and instead deploy the subscription template directly, but using the DeploymentSubscription parameter.
> 
> PS: [#137 (comment)](https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2372453056) CLI: [#137 (comment)](https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2067033907)
> 
> So the cmd for [@pbo-top](https://github.com/pbo-top) would be something like the following (adding a line break and rearranging parameters for clarity)
> 
> ```
> az stack mg create `
> --name 'test-budget-stack-001' `
> --location 'swedencentral' `
> --management-group-id 'deploy-stack-test-001' `
> --parameters subscriptionID='234234-23234...`
> --action-on-unmanage 'deleteResources' `
> --deny-settings-mode 'denyWriteAndDelete' `
> --deny-settings-excluded-principals '1231-123123-123123...' `
> --deny-settings-apply-to-child-scopes
> 
> --template-file 'budget1.bicep' ` <<< ***CHANGED***
> --deployment-subscription 248248-....  <<< ***ADDED***
> ```

You're right , that was the thing I was missing, I kept trying to use the "wrapper" bicep file
Thank you

**Comment by pbo-top** [comment]
> [@pbo-top](https://github.com/pbo-top) [@MichalZezula](https://github.com/MichalZezula) Deployment Stacks absolutely supports this scenario!
> 
@snarkywolverine  Thanks for the suggestion, yes ditching the wrapper made the deployment go through.

Now, looking at UI the Stack seems OK with the deny-setting:
![Image](https://github.com/user-attachments/assets/b625db48-3186-4103-872a-5fc0cbecaf15)

But... this leaves me with the same problem that @paul-mccormack mentioned previously with policies. Users in the target Subscription who eg. have the IAM Contributer role can modify or delete the budget (the user has no rights at the MGMT levels). The ManagementGroup stacks' deny does not seem to have any effect at the Subscription level.

**Comment by snarkywolverine** [comment]
@pbo-top Can you elaborate on this: _Users in the target Subscription who eg. have the IAM Contributer role can modify or delete the budget_.

Full disclosure, I have not worked much with budgets - are they updateable outside of the Azure control plane? The deny assignments are only going to help within the context of Azure RBAC.

If a resource is updatable/deletable through Azure despite a deny assignment, I would appreciate a correlation ID on the update/delete call - that would indicate the deny assignment is not being respected for some reason.

The expectation from our side is that the deny assignment should protect the resource, and users shouldn't be able to modify the stack to give themselves permission (since the stack lives at a different scope where users don't have permission).

**Comment by pbo-top** [comment]
@snarkywolverine 

I don't know what MS is doing behind lines, aka. if Budgets can be seen as resource at the same level as stuff one can place in a ResourceGroup, but the Budgets are governed by the IAM RBAC and can be deployed through the RM.

What I've tested is creating a Budget via a "mg" stack deployment as suggested above. This ends up as the screenshot I pasted, which is the stack seen after selecting the Management Group (named "Deploy Stack Test 001"), then selecting "Deployment Stacks".

I've made two test regarding the correlationId,

1. Deploy the MGMT stack, then in a new private browser logon with a test-user who is Contributer on the Subscription and nothing else. He can modify and delete the Budget. "correlationId": "e21ff930-0639-44b1-ba0c-00179557dbf1", "eventTimestamp": "2024-11-06T10:13:33.0989466Z" (CET).

2. Deploy the MGMT stack, then in a new private browser logon with a test-user who have IAM/RBAC "Cost Management Contributor" and nothing else. He can modify and delete the Budget. "correlationId": "997bded0-f233-4fc1-98b9-08cbc81e54c8", "eventTimestamp": "2024-11-06T11:17:08.1023723Z" (CET).

The last test was to see if there was some especially funky with the higher privileged roles as Contributer or Owner, but it does appear so. Everybody who have an IAM/RBAC to fiddle with the Budget can do so, regardless of the deny's in the stack.

**Comment by snarkywolverine** [comment]
@pbo-top Sorry for the delay in responding. I am working with some of our partner teams to investigate the issues related to Microsoft.Consumption/budget.

Meanwhile, I have created #196 to track those issues -- the issue with Budget resources is different than the creation of deny assignments on an MG-scoped stack.

**Comment by DennisSchouwenaars** [comment]
I've been running several tests today for the same scenario as well. And the conclusion I made in regards to this issue is that the entire "Deny Assignment" isn't even created on the Subscription level at all. If I have f.e. resourceGroups with child resources in the same DeploymentStack I notice that they do have a "Deny Assignment" visible on the IAM tab. I haven't tested yet if this changes if I add additional Subscription based resources.

**Comment by willemsen-it** [comment]
I also need this to work because I am using the 'Azure Landing Zone Vending Machine' from Microsoft and the vending machine uses multiple scopes in it's modules (mainly managementGroup and subscription). It needs to be able to deploy subscriptions and assign those to the right managementGroup and it needs to deploy resources in the subscriptions. I want the stack to be created at the managementgroup level (so no owner of a subscription a.k.a. customer can delete or manage the stack). I also want to protect the resources that are deployed through a deny setting.

The problem we have now is that I cannot use the Deny settings because the code contains the scope 'managementGroup' in the files: managementGroupSubscription.bicep, subResourceWrapper.bicep and subscriptionAlias.bicep.
When I try to create the deployment stack it states "deny settings are only allowed for resources deployed to subscription level not management group level" but I also cannot use the parameter "--deployment-subscription <sub id>" because the vending machine contains files/modules for the subscription deployment at managementgroup scope.

**Comment by dominikheeb** [comment]
Hi, I am also upvoting this issue. We want to deploy platform resources, for example spoke networks to the landing zone subscriptions, but allow the landing zone owners to get the owner role on their subscription. This Feature would allow us to manage the resources in our deployment stack in the management group for the landing zones and keep the role concept simple in the landing zone subscription

**Comment by snarkywolverine** [comment]
@dominikheeb Will the topology discussed above work for your scenarios?

https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2455399003

In short, you should be able to create the deployment stack at the mgmt group scope, but specify a target subscription to deploy the underlying resources.

**Comment by snarkywolverine** [comment]
@willemsen-it Thanks for the feedback. Conceptually, if you were able to deploy the landing zone with a stack that protected the subscription resources (but not the mg-scoped resources), would that work for your needs?  Or do you need the deny assignments to live on mgmt group scoped resources created by the LZ as well?

**Comment by willemsen-it** [comment]
@snarkywolverine yes that would work. If I can create the deploymentstack at managementgroup level and protect the resources that are deployed at subscription level with a deny assignment then that would be great.

**Comment by dominikheeb** [comment]
> [@dominikheeb](https://github.com/dominikheeb) Will the topology discussed above work for your scenarios?
> 
> [#137 (comment)](https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2455399003)
> 
> In short, you should be able to create the deployment stack at the mgmt group scope, but specify a target subscription to deploy the underlying resources.

Thanks for the answer, but no.
This doesn't work, since when i use for example a deployment stack to create the resources in my management subscription and the connectivity subscription, the deny policy is only applied to the resources in the management subscription.
The resources from the connectivity subscription are listed but for the deny policy there is written inapplicable.

If we now want to deploy the resources in the different landing zones its only possible when we create a deployment stack in each of the subscriptions... but this would defeat the purpose in my opinion, since we want to give out ownership to the landing zone into the projects, but if they can just delete the stack this doesnt really make sense

**Comment by maikvandergaag** [comment]
> > [@dominikheeb](https://github.com/dominikheeb) Will the topology discussed above work for your scenarios?
> > [#137 (comment)](https://github.com/Azure/deployment-stacks/issues/137#issuecomment-2455399003)
> > In short, you should be able to create the deployment stack at the mgmt group scope, but specify a target subscription to deploy the underlying resources.
> 
> Thanks for the answer, but no. This doesn't work, since when i use for example a deployment stack to create the resources in my management subscription and the connectivity subscription, the deny policy is only applied to the resources in the management subscription. The resources from the connectivity subscription are listed but for the deny policy there is written inapplicable.
> 
> If we now want to deploy the resources in the different landing zones its only possible when we create a deployment stack in each of the subscriptions... but this would defeat the purpose in my opinion, since we want to give out ownership to the landing zone into the projects, but if they can just delete the stack this doesnt really make sense

We also see this issue. This especially happens when deploying resources to multiple subscriptions and you want to manage them in a stack. Happy to share testing details when needed.
